core/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java

```diff
@@ -816,6 +816,12 @@ public boolean append(Object kbase, long koff, int klen, Object vbase, long voff
           } catch (SparkOutOfMemoryError oom) {
             canGrowArray = false;
           }
+        } else if (numKeys >= growthThreshold && longArray.size() / 2 >= MAX_CAPACITY) {
```

Comment:
it should be more readable to combine the common condition:
```
// The map has reached its growth threshold, try to grow it
if (numKeys >= growthThreshold) {
  // We use two array entries per key ...
  if (longArray.size() / 2 < MAX_CAPACITY) {
     ...
  } else {
    // It's already at MAX_CAPACITY and cannot grow ...
    canGrowArray = false;
  }
}
```

Reply:
+1 for @cloud-fan 's refactoring suggestion.

Reply:
+1 too

Reply:
Thanks, done.